common/common.go

```diff
@@ -553,31 +553,31 @@ func ExcludeError(err, excl error) error {
 }
 
 // ErrorCollector allows collecting a stream of errors from concurrent go routines
-// Users should call e.Wg.Done and send errors to e.C
 type ErrorCollector struct {
-	C  chan error
-	Wg sync.WaitGroup
+	errs error
+	wg   sync.WaitGroup
+	m    sync.Mutex
 }
 
-// CollectErrors returns an ErrorCollector with WaitGroup and Channel buffer set to n
-func CollectErrors(n int) *ErrorCollector {
-	e := &ErrorCollector{
-		C: make(chan error, n),
-	}
-	e.Wg.Add(n)
-	return e
+// Collect waits for the internal wait group to be done and returns an error collection
+func (e *ErrorCollector) Collect() (errs error) {
+	e.wg.Wait()
+	e.m.Lock()
+	defer e.m.Unlock()
+	return e.errs
 }
 
-// Collect runs waits for e.Wg to be Done, closes the error channel, and return a error collection
-func (e *ErrorCollector) Collect() (errs error) {
-	e.Wg.Wait()
-	close(e.C)
-	for err := range e.C {
-		if err != nil {
-			errs = AppendError(errs, err)
-		}
+// Go runs a function in a goroutine and collects any error it returns
+func (e *ErrorCollector) Go(f func() error) {
+	if f == nil {
+		panic("ErrorCollector.Go received a nil function")
```

Comment:
I would argue that if things are the caller's responsibility, which is something you often advocate, then this should just return rather than panic. A user who is calling this can check if the function that they are passing in is nil before doing so and can panic on their own terms

Reply:
It's usually a closure right? Even in sync.Waitgroup `Go` func call it panics if func is nil. I only just added this check outside of the lower go routine for easier debugging.

Reply:
It does not panic on nil. If it did I'd argue to not introduce an extra one, but much less strongly. I don't know what you mean about closure though.

This isn't something I'm very fussed on, but I don't see the need to introduce it